storm-client/test/jvm/org/apache/storm/windowing/persistence/WindowStateTest.java

```diff
@@ -0,0 +1,247 @@
+
+        Mockito.when(supplier.get()).thenReturn(Collections.emptyMap());
+
+        WindowState<Integer> ws = getWindowState(10 * MAX_EVENTS_PER_PARTITION);
+
+        long partitions = 15;
+
+        long numEvents = partitions * MAX_EVENTS_PER_PARTITION;
+        List<Event<Integer>> expected = new ArrayList<>();
+        for (int i = 0; i < numEvents; i++) {
+            Event<Integer> event = getEvent(i);
+            expected.add(event);
+            ws.add(event);
+        }
+
+        Assert.assertEquals(5, partitionMap.size());
+        Iterator<Event<Integer>> it = ws.iterator();
+        List<Event<Integer>> actual = new ArrayList<>();
+        it.forEachRemaining(actual::add);
+        Assert.assertEquals(expected, actual);
+
+        // iterate again
+        it = ws.iterator();
+        actual.clear();
+        it.forEachRemaining(actual::add);
+        Assert.assertEquals(expected, actual);
+
+        // remove
+        it = ws.iterator();
+        while (it.hasNext()) {
+            it.next();
+            it.remove();
+        }
+
+        it = ws.iterator();
+        actual.clear();
+        it.forEachRemaining(actual::add);
+        Assert.assertEquals(Collections.emptyList(), actual);
+    }
+
+    @Test
+    public void testIteratorPartitionNotEvicted() throws Exception {
+        Map<Long, WindowState.WindowPartition<Event<Tuple>>> partitionMap = new HashMap<>();
+        Mockito.when(partitionIdsState.get(Mockito.any(), Mockito.any())).then(returnsArgAt(1));
+        Mockito.when(windowState.get(Mockito.any(), Mockito.any())).then(new Answer<Object>() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                Object[] args = invocation.getArguments();
+                WindowState.WindowPartition<Event<Tuple>> evicted = partitionMap.get(args[0]);
+                return evicted != null ? evicted : args[1];
+            }
+        });
+
+        Mockito.doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                Object[] args = invocation.getArguments();
+                partitionMap.put((long)args[0], (WindowState.WindowPartition<Event<Tuple>>)args[1]);
+                return null;
+            }
+        }).when(windowState).put(Mockito.any(), Mockito.any());
+
+        Mockito.when(supplier.get()).thenReturn(Collections.emptyMap());
+
+        WindowState<Integer> ws = getWindowState(10 * MAX_EVENTS_PER_PARTITION);
+
+        long partitions = 10;
+
+        long numEvents = partitions * MAX_EVENTS_PER_PARTITION;
+        List<Event<Integer>> expected = new ArrayList<>();
+        for (int i = 0; i < numEvents; i++) {
+            Event<Integer> event = getEvent(i);
+            expected.add(event);
+            ws.add(event);
+        }
+
+        Iterator<Event<Integer>> it = ws.iterator();
+        for(int i=0; i<9500; i++) {
```

Comment:
I think it would be helpful to put a note here about this causing the partition to be pinned. At a glance it's not obvious why it should work this way.